File: scripts/parity/schema/ruby/dump.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Usage (from repo root):
#   bundle exec --gemfile scripts/parity/schema/ruby/Gemfile \
#     ruby scripts/parity/schema/ruby/dump.rb <fixture-dir> <out.json>
#
# fixture-dir and out.json paths can be relative or absolute.
#
# Applies <fixture-dir>/schema.sql to a fresh SQLite database, introspects
# it using ActiveRecord, canonicalizes the result, and writes canonical JSON
# to <out.json>.
#
# Validates against <fixture-dir>/expected.json (D6) and exits 2 on mismatch.

require "bundler/setup"
require "active_record"
require "sqlite3"
require "tmpdir"
require "json"
require "fileutils"
require_relative "canonicalize"

FILTERED_TABLES = %w[schema_migrations ar_internal_metadata].freeze

def usage
  warn "Usage: bundle exec ruby dump.rb <fixture-dir> <out.json>"
  exit 1
end

fixture_dir, out_path = ARGV
usage unless fixture_dir && out_path

fixture_dir = File.expand_path(fixture_dir)
out_path    = File.expand_path(out_path)

Dir.mktmpdir("parity-ruby-") do |tmpdir|
  db_path = File.join(tmpdir, "schema.db")

  begin
    # 1. Apply schema.sql to a fresh temp SQLite file
    sql = File.read(File.join(fixture_dir, "schema.sql"))
    SQLite3::Database.new(db_path) do |db|
      db.execute_batch(sql)
    end

    # 2. Connect via ActiveRecord
    ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: db_path)
    conn = ActiveRecord::Base.connection

    # 3. Introspect tables, columns, indexes, PKs
    tables = conn.tables.reject { |t| FILTERED_TABLES.include?(t) }.sort

    native_dump = {}
    tables.each do |table_name|
      cols = conn.columns(table_name).map do |col|
        {
          name:      col.name,
          ar_type:   col.type,
          null:      col.null,
          default:   col.default,
          limit:     col.limit,
          precision: col.precision,
          scale:     col.scale,
        }
      end

      raw_pk = conn.primary_key(table_name)
      pk_cols = case raw_pk
      when nil    then []
      when Array  then raw_pk
      else             [raw_pk]
      end

      idxs = conn.indexes(table_name).map do |idx|
        {
          name:    idx.name,
          columns: idx.columns,
          unique:  idx.unique,
          where:   idx.where,
        }
      end

      native_dump[table_name] = {
        columns:             cols,
        indexes:             idxs,
        primary_key_columns: pk_cols,
      }
    end

    # 4. Canonicalize
    canonical = Canonicalize.call(native_dump)

    # 5. Validate against expected.json (D6)
    expected = JSON.parse(File.read(File.join(fixture_dir, "expected.json")))

    actual_tables   = canonical["tables"].map { |t| t["name"] }.sort
    expected_tables = expected["tables"].sort
    if actual_tables != expected_tables
      warn "parity dump: table mismatch\n  expected: #{expected_tables.inspect}\n  actual:   #{actual_tables.inspect}"
      exit 2
    end

    actual_index_count = canonical["tables"].sum { |t| t["indexes"].length }
    if actual_index_count != expected["indexCount"]
      warn "parity dump: index count mismatch\n  expected: #{expected["indexCount"]}\n  actual:   #{actual_index_count}"
      exit 2
    end

    # 6. Write canonical JSON
    FileUtils.mkdir_p(File.dirname(out_path))
    File.write(out_path, JSON.pretty_generate(canonical) + "\n")
    puts "parity dump (rails): wrote #{out_path}"

  ensure
    begin
      ActiveRecord::Base.remove_connection
    rescue StandardError
      # already removed or never opened
    end
  end
end

File: scripts/parity/schema/node/dump.ts
#!/usr/bin/env tsx
/**
 * Usage: tsx scripts/parity/schema/node/dump.ts <fixture-dir> <out.json>
 *
 * Must be run from the repo root so fixture paths and output paths resolve correctly.
 *
 * Applies <fixture-dir>/schema.sql to a fresh SQLite database, introspects
 * it using the trails ActiveRecord adapter, canonicalizes the result, and
 * writes canonical JSON to <out.json>.
 *
 * Validates against <fixture-dir>/expected.json (D6) and exits 2 on mismatch.
 */

import Database from "better-sqlite3";
import { readFileSync, mkdtempSync, writeFileSync, rmSync, existsSync } from "node:fs";
import { tmpdir } from "node:os";
import { join, resolve } from "node:path";
// Relative imports so tsx can resolve from source without a prior build.
// These paths are resolved relative to this file's location by the module system.
import { Base } from "../../../../packages/activerecord/src/base.js";
import {
  introspectTables,
  introspectColumns,
  introspectIndexes,
  introspectPrimaryKey,
} from "../../../../packages/activerecord/src/schema-introspection.js";
import { canonicalize } from "./canonicalize.js";
import type { NativeDump, NativeColumn, NativeIndex } from "./canonicalize.js";

interface ExpectedManifest {
  tables: string[];
  indexCount: number;
}

function usage(): never {
  process.stderr.write("Usage: tsx scripts/parity/schema/node/dump.ts <fixture-dir> <out.json>\n");
  process.exit(1);
}

function assertRepoRoot(): void {
  // Fixture arguments (argv[2], argv[3]) are resolved relative to CWD.
  // Fail fast with a clear message rather than a cryptic path error later.
  if (!existsSync("packages/activerecord/src/base.ts")) {
    process.stderr.write(
      "parity dump: must be run from the repo root (packages/activerecord/src/base.ts not found)\n",
    );
    process.exit(1);
  }
}

const FILTERED_TABLES = new Set(["schema_migrations", "ar_internal_metadata"]);

async function main(): Promise<void> {
  assertRepoRoot();
  const [fixtureDir, outPath] = process.argv.slice(2);
  if (!fixtureDir || !outPath) usage();

  const fixtureDirAbs = resolve(fixtureDir);
  const outPathAbs = resolve(outPath);

  const tmpDir = mkdtempSync(join(tmpdir(), "parity-node-"));
  const dbPath = join(tmpDir, "schema.db");

  try {
    // 1. Apply schema.sql to a fresh temp SQLite file via better-sqlite3
    const sql = readFileSync(join(fixtureDirAbs, "schema.sql"), "utf8");
    const db = new Database(dbPath);
    try {
      db.exec(sql);
    } finally {
      db.close();
    }

    // 2. Connect via trails adapter
    // Pass dbPath directly — adapterNameFromUrl recognises .db extension as sqlite.
    await Base.establishConnection(dbPath);
    const adapter = Base.adapter;

    // 3. Introspect tables, columns, indexes
    const tables = (await introspectTables(adapter)).filter((t) => !FILTERED_TABLES.has(t)).sort();

    const nativeDump: NativeDump = {};

    for (const tableName of tables) {
      const cols = await introspectColumns(adapter, tableName);
      const idxDefs = await introspectIndexes(adapter, tableName);

      const columns: NativeColumn[] = cols.map((col) => ({
        name: col.name,
        sqlType: col.sqlType ?? col.type ?? "",
        primaryKey: col.primaryKey,
        null: col.null,
        default: col.default !== null && col.default !== undefined ? String(col.default) : null,
        limit: col.limit,
        precision: col.precision,
        scale: col.scale,
      }));

      const indexes: NativeIndex[] = idxDefs.map((idx) => {
        if (!idx.name || idx.name.trim() === "") {
          throw new Error(
            `parity dump: index on "${tableName}" has no name (columns: ${JSON.stringify(idx.columns)})`,
          );
        }
        return {
          name: idx.name,
          columns: idx.columns,
          unique: idx.unique,
          where: idx.where ?? null,
        };
      });

      const primaryKeyColumns = await introspectPrimaryKey(adapter, tableName);
      nativeDump[tableName] = { columns, indexes, primaryKeyColumns };
    }

    // 4. Canonicalize
    const canonical = canonicalize(nativeDump);

    // 5. Validate against expected.json (D6)
    const expected = JSON.parse(
      readFileSync(join(fixtureDirAbs, "expected.json"), "utf8"),
    ) as ExpectedManifest;

    const actualTableNames = canonical.tables.map((t) => t.name).sort();
    const expectedTableNames = [...expected.tables].sort();
    if (JSON.stringify(actualTableNames) !== JSON.stringify(expectedTableNames)) {
      process.stderr.write(
        `parity dump: table mismatch\n  expected: ${JSON.stringify(expectedTableNames)}\n  actual:   ${JSON.stringify(actualTableNames)}\n`,
      );
      process.exitCode = 2;
      return;
    }

    const actualIndexCount = canonical.tables.reduce((n, t) => n + t.indexes.length, 0);
    if (actualIndexCount !== expected.indexCount) {
      process.stderr.write(
        `parity dump: index count mismatch\n  expected: ${expected.indexCount}\n  actual:   ${actualIndexCount}\n`,
      );
      process.exitCode = 2;
      return;
    }

    // 6. Write canonical JSON
    writeFileSync(outPathAbs, JSON.stringify(canonical, null, 2) + "\n");
    process.stdout.write(`parity dump (trails): wrote ${outPathAbs}\n`);
  } finally {
    // Explicitly close the adapter's DB handle before removing the temp file.
    // Base.removeConnection() only drops the pool reference; the adapter (and
    // its better-sqlite3 handle) may still be open, causing EBUSY on Windows.
    try {
      const a = Base.adapter as { close?: () => void };
      if (typeof a.close === "function") a.close();
    } catch {
      /* adapter unavailable or already closed */
    }
    try {
      Base.removeConnection();
    } catch {
      /* already removed or never opened */
    }
    try {
      rmSync(tmpDir, { recursive: true, force: true });
    } catch (err) {
      process.stderr.write(
        `parity dump: warning: failed to remove temp dir ${tmpDir}: ${err instanceof Error ? err.message : String(err)}\n`,
      );
    }
  }
}

main().catch((err: unknown) => {
  process.stderr.write(`parity dump: ${err instanceof Error ? err.message : String(err)}\n`);
  process.exit(1);
});
